payment-failure/src/main/scala/com/gu/identity/paymentfailure/IdentityClient.scala

```diff
@@ -1,41 +1,64 @@
 package com.gu.identity.paymentfailure
 
+import cats.syntax.either._
 import com.typesafe.scalalogging.StrictLogging
 import io.circe.{Decoder, Encoder}
-import io.circe.generic.semiauto._
+import io.circe.generic.JsonCodec
+import io.circe.parser.decode
 import io.circe.syntax._
 import scalaj.http._
 
-case class IdentityConfig(identityApiHost: String, clientAccessToken: String)
+class IdentityClient(config: Config) extends StrictLogging {
+  import IdentityClient._
 
-case class IdentityEmailTokenRequest(email: String)
+  // Given a way of serialising the request body (Req) as JSON
+  // and deserialising the response body to the target case class (Res)
+  // execute a POST request against the given path.
+  private def executePostRequest[Req: Encoder, Res: Decoder](path: String, requestBody: Req): Either[Throwable, Res] = {
+    logger.info(s"executing POST $path")
 
-object IdentityEmailTokenRequest {
-  implicit val identityEmailTokenEncoder: Encoder[IdentityEmailTokenRequest] = deriveEncoder[IdentityEmailTokenRequest]
-}
+    val request = Http(s"${config.idapiHost}$path")
+      .postData(requestBody.asJson.noSpaces)
+      .header("X-GU-ID-Client-Access-Token", s"Bearer ${config.idapiAccessToken}")
+      .header("content-type", "application/json")
 
-case class IdentityEmailTokenResponse(status: String, encryptedEmail: String)
+    Either.catchNonFatal(request.asString)
+      .leftMap[Throwable](err => networkError(path, err))
+      .flatMap {
+        case res if res.is2xx => decode[Res](res.body).leftMap(err => decodingError(path, err))
+        case res => Left(non2xxError(path, res.code, res.body))
+      }
+  }
 
-object IdentityEmailTokenResponse {
-  implicit val identityEmailTokenDecoder: Decoder[IdentityEmailTokenResponse] = deriveDecoder[IdentityEmailTokenResponse]
+  def encryptEmail(requestBody: IdentityEmailTokenRequest): Either[Throwable, IdentityEmailTokenResponse] =
+    executePostRequest[IdentityEmailTokenRequest, IdentityEmailTokenResponse]("/signin-token/email", requestBody)
+
+  def createAutoSignInToken(requestBody: AutoSignInLinkRequestBody): Either[Throwable, AutoSignInLinkResponseBody] =
+    executePostRequest[AutoSignInLinkRequestBody, AutoSignInLinkResponseBody]("/magic-link/generate", requestBody)
 }
 
-class IdentityClient(config: Config) extends StrictLogging {
+object IdentityClient {
```

Comment:
Created a companion object to hold methods and classes directly related to the `IdentityClient` class